dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -1452,4 +1584,47 @@ private long countMaySet() {
     private long countMustSet() {
         return knowledgeMap.values().stream().mapToLong(k -> k.must.size()).sum();
     }
+
+    // This models same_location_r from the PTX Alloy model.
+    // Virtual addresses always alias to a physical one.
+    // Physical addresses may alias to other physical one.
+    // There are 4 cases (-> means alias):
+    // (1) virtual/virtual
+    //     `x -> px -> py <- y` 
+    //     `x -> px <- py <- y` 
+    // (2,3) physical/virtual
+    //     `x <- py <- y` 
+    //     `x -> px <- py <- y` (x is physical but still alias to physical px) 
+    // (4) physical/physical
+    //     `x <- y` 
+    //     `x -> y` 
+
+    private boolean alias(MemEvent e1, MemEvent e2) {
+        // TODO: Add support for pointers
+        if (!(e1.getAddress() instanceof MemoryObject) || !(e2.getAddress() instanceof MemoryObject)) {
+            return false;
+        }
+        MemoryObject x = (MemoryObject) e1.getAddress();
+        MemoryObject y = (MemoryObject) e2.getAddress();
+        boolean virtual1 = x.isVirtual();
+        boolean virtual2 = y.isVirtual();
+        MemoryObject px = x.getAlias();
+        MemoryObject py = y.getAlias();
+        if (virtual1 && virtual2) {
+            // Virtual addresses always alias to a physical one
+            assert (px != null);
+            assert (py != null);
+            return py.equals(px.getAlias()) || px.equals(py.getAlias());
+        } else if (!virtual1 && virtual2) {
+            // Virtual addresses always alias to a physical one
+            assert (py != null);
+            return x.equals(py.getAlias()) || py.equals(px);
+        } else if (virtual1) { // virtual1 && !virtual2
+            // Virtual addresses always alias to a physical one
+            assert (px != null);
+            return y.equals(px.getAlias()) || px.equals(py);
+        } else { // !virtual1 && !virtual2
+            return x.equals(py) || y.equals(px);
+        }
+    }
```

Comment:
This code is very similar to `AliasAnalysis.virtualAlias`. 
Can this be somehow unified?
Also, the method name is bad. Checks like `alias.mayAlias(e1 ,e2) || alias(e1, e2)` look awkward and hard to understand.

Reply:
This function has been rewritten as `samePhysicalAddress`